jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/VersionChecker.java

```diff
@@ -0,0 +1,143 @@
+ *       akin to {@code [1.0,)}
+ * </ol>
+ *
+ * To support custom version representations, the actual version object type ({@code <V>}) is
+ * pluggable. It must implement {@link Comparable}. The versions in the range must have at most 3
+ * components (e.g., {@code major.minor.micro}).
+ *
+ * <p>This class exists as Gradle has no version-range-type class and Maven's {@code
+ * org.apache.maven.artifact.versioning.VersionRange} treats a single version as an exact bound.
+ * Note that Gradle's {@code org.gradle.util.GradleVersion} class does not support
+ * major-version-only versions.
+ */
+public class VersionChecker<V extends Comparable<? super V>> {
+  /** The expected version representation. */
+  private static final String VERSION_PATTERN = "\\d+(\\.\\d+(\\.\\d+)?)?";
+
+  // Helper functions to avoid the cognitive burden of {@link Comparable#compareTo()}
+
+  /** Return {@code true} if {@code a} is less than {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean LT(T a, T b) {
+    return a.compareTo(b) < 0;
+  }
+
+  /** Return {@code true} if {@code a} is less than or equal to {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean LE(T a, T b) {
+    return a.compareTo(b) <= 0;
+  }
+
+  /** Return {@code true} if {@code a} is greater than {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean GT(T a, T b) {
+    return a.compareTo(b) > 0;
+  }
+
+  /** Return {@code true} if {@code a} is greater than or equal to {@code b}. */
+  @VisibleForTesting
+  static <T extends Comparable<? super T>> boolean GE(T a, T b) {
+    return a.compareTo(b) >= 0;
+  }
+
+  /** Responsible for converting a string representation to a comparable version representation. */
+  private Function<String, V> converter;
+
+  public VersionChecker(Function<String, V> converter) {
+    this.converter = converter;
+  }
+
+  /**
+   * Return {@code true} if {@code actualVersion} is contained within the version range represented
+   * by {@code acceptableVersionRange}.
+   *
+   * @param acceptableVersionRange the encoded version range
+   * @param actualVersion the version to be compared
+   * @return true if the version is acceptable
+   * @throws IllegalArgumentException if the version could not be parsed
+   */
+  @VisibleForTesting
+  public boolean compatibleVersion(String acceptableVersionRange, String actualVersion) {
+    V pluginVersion = parseVersion(actualVersion);
+
+    // Treat a single version "1.4" as a left bound, equivalent to "[1.4,)"
+    if (acceptableVersionRange.matches(VERSION_PATTERN)) {
+      return GE(pluginVersion, parseVersion(acceptableVersionRange));
+    }
+
+    // Otherwise ensure it is a version range with bounds
+    Preconditions.checkArgument(
+        acceptableVersionRange.matches(
+            "[\\[(](" + VERSION_PATTERN + ")?,(" + VERSION_PATTERN + ")?[\\])]"),
```

Comment:
Usually escaping is unnecessary inside `[ ... ]` (for example `.` means a period inside it too), so `[[(]` might work? Same for the closing pattern.

Reply:
A backslash is required for the pattern matching the opening bracket/paren, and not required for the trailing bracket/paren `[])]`.

Reply:
OK, never mind then. Let's not touch the pattern anymore, since escaping eliminates ambiguity.